spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/factory/RetryGatewayFilterFactory.java

```diff
@@ -182,7 +233,14 @@ public GatewayFilter apply(Repeat<ServerWebExchange> repeat,
 						.repeatWhen(repeat.withApplicationContext(exchange));
 			}
 
-			return Mono.fromDirect(publisher);
+			return Mono.fromDirect(publisher).doFinally(signalType -> {
+				Optional<DataBuffer> originalRequestBuffer = exchange
+						.getAttributeOrDefault(RETRY_BODY_KEY, Optional.empty());
+				// If we have a buffer here then attempt to release it
+				if (originalRequestBuffer.isPresent()) {
+					DataBufferUtils.release(originalRequestBuffer.get());
```

Comment:
We might have to loop this here while the buffer is allocated.  I suppose it is possible that we replay and retry.  In that case we want to release until we are not allocated.